pkg/dispers/dispers/requests.go

```diff
@@ -27,3 +32,178 @@ type InputCI struct {
+	Type      NodeType    `json:"type"`
+	Value     string      `json:"value,omitempty"`
+	LeftNode  interface{} `json:"left_node,omitempty"`
+	RightNode interface{} `json:"right_node,omitempty"`
+}
+
+// Compute compute the OperationTree and returns the list of encrypted addresses
+func (o *OperationTree) Compute(listsOfAddresses map[string][]string) ([]string, error) {
+
+	if o.Type == SingleNode {
+		// Retrieve list of addresses from listsOfAddresses
+		val, ok := listsOfAddresses[o.Value]
+		if !ok {
+			return []string{}, errors.New("Unknown concept")
+		}
+		return val, nil
+
+	} else if o.Type == UnionNode || o.Type == IntersectionNode {
+
+		// Compute operations on LeftNode and RightNode
+		leftNode := o.LeftNode.(OperationTree)
+		a, err := leftNode.Compute(listsOfAddresses)
+		if err != nil {
+			return []string{}, err
+		}
+		rightNode := o.RightNode.(OperationTree)
+		b, err := rightNode.Compute(listsOfAddresses)
+		if err != nil {
+			return []string{}, err
+		}
+		// Compute operation between LeftNode and RightNode
+		switch o.Type {
+		case UnionNode:
+			return union(a, b), nil
+		case IntersectionNode:
+			return intersection(a, b), nil
+		default:
+			return []string{}, errors.New("Unknown type")
+		}
+	} else {
+		return []string{}, errors.New("Unknown type")
+	}
+}
+
+// UnmarshalJSON is used to load the OperationTree given by the Querier
+func (o *OperationTree) UnmarshalJSON(data []byte) error {
+
+	var v map[string]interface{}
+	if err := json.Unmarshal(data, &v); err != nil {
+		return err
+	}
+
+	// Retrieve the NodeType
+	if v["type"] == nil {
+		return errors.New("No type defined")
+	}
+	switch int(v["type"].(float64)) {
+	case 0:
+		o.Type = SingleNode
+	case 1:
+		o.Type = UnionNode
+	case 2:
+		o.Type = IntersectionNode
+	default:
+		return errors.New("Unknown type")
+	}
+
+	// Retrieve others attributes depending on NodeType
+	switch {
+	case o.Type == SingleNode:
+		o.Value, _ = v["value"].(string)
+	case o.Type == IntersectionNode || o.Type == UnionNode:
+		var leftNode OperationTree
+		var rightNode OperationTree
+
+		leftNodeByte, _ := json.Marshal(v["left_node"].(map[string]interface{}))
+		err := json.Unmarshal(leftNodeByte, &leftNode)
+		if err != nil {
+			return err
+		}
+
+		rightNodeByte, _ := json.Marshal(v["right_node"].(map[string]interface{}))
+		err = json.Unmarshal(rightNodeByte, &rightNode)
+		if err != nil {
+			return err
+		}
+		o.LeftNode = leftNode
+		o.RightNode = rightNode
+	default:
+	}
+	return nil
+}
+
+// InputTF contains a map that associate every concept to a list of Addresses
+// and a operation to compute to retrive the final list
+type InputTF struct {
+	IsEncrypted               bool                `json:"isencrypted,omitempty"`
+	EncryptedListsOfAddresses []byte              `json:"enc_instances,omitempty"`
+	EncryptedTargetProfile    []byte              `json:"enc_operation,omitempty"`
+	ListsOfAddresses          map[string][]string `json:"instances,omitempty"`
```

Comment:
I am still unsure of why not just `[]Instance` as the type ?

Reply:
1. TF should not read Instance object because TF sould not have access to that kind of info. TF will manipulate strings as encrypted-marshal of Instances
2. And why a map ? Because concepts are pseudo-anonymised from the querier. This new name is used to define the target profile. And ListsOfAddresses associate that new name to the list of instances, that is in fact a list of strings.

Maybe I could transform ListOfAddresses in OutputTF to type []string ! It will be better !

Reply:
Ok, it was a bit of a dumb question, but it is clearer now